### PR TITLE
Lock devices during OVF export as this could cause a race otherwise

### PR DESCRIPTION
**Pull request by dupondje**
Closes: #797

During the creation of the ovf config export, oVirt does modifications to the VmDevices.
Now another thread cleans those entries again, and if you are unlucky it happens during the ovf generation itself.

DEBUG [org.ovirt.engine.core.vdsbroker.monitoring.VmDevicesMonitoring] (EE-ManagedScheduledExecutorService-engineScheduledThreadPool-Thread-60) [] VM 'xxxx' unmanaged device was marked for remove : {1}

And this causes the export to be skipped.

So we add a device lock during the ovf export to resolve this.

Signed-off-by: Jean-Louis Dupond <jean-louis@dupond.be>

### CODE REVIEW

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/GetVmOvfByVmIdQuery.java [hunk 1]
```diff
@@ -35,10 +39,16 @@ protected void executeQueryCommand() {
             return;
         }
 
+        // We want to lock the devices to make sure no other thread modifies them during the ovf export.
+        // Otherwise we end up without ovf in the export
+        resourceManager.getVmManager(getParameters().getId()).getVmDevicesLock().lock();
```
Comment (ahadas) [inline backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/GetVmOvfByVmIdQuery.java:44]: that won't cover all cases - the db generation may change due to an operation (and not via the devices monitoring) and you also may get `vm.getDbGeneration() != getParameters().getRequiredGeneration()` above

**Review by ahadas — COMMENTED** [review]
it doesn't sound like a good approach to block the monitoring because of a query - I'd rather suggest to let the client re-invoke the query if it requests an OVF and doesn't get one

**Comment by dupondje** [comment]
True, but now we are racing against the VmDevicesMonitoring to get the ovf info.
We currently request the ovf data of around 400 vm's and like 10-20 are missing the ovf info. So its not it only happens in some edge cases :)

**Comment by ahadas** [comment]
> True, but now we are racing against the VmDevicesMonitoring to get the ovf info. We currently request the ovf data of around 400 vm's and like 10-20 are missing the ovf info. So its not it only happens in some edge cases :)

10-20 out of 400 is more than I'd expect, yes
do you do many hot-plugs or snapshots?

**Comment by dupondje** [comment]
> > True, but now we are racing against the VmDevicesMonitoring to get the ovf info. We currently request the ovf data of around 400 vm's and like 10-20 are missing the ovf info. So its not it only happens in some edge cases :)
> 
> 10-20 out of 400 is more than I'd expect, yes do you do many updates / hot-plugs / snapshots?

Well at the moment when we do the API call there are no changes done. So its purely because the VmDevicesMonitoring is removing the unmanaged devices to quickly when the ovf exporter just added them to generate the ovf.

I did not check the whole ovf export code why it exactly needs to add unmanaged devices to the vm to take the VM export.
But it seems to be suboptimal.

**Comment by ahadas** [comment]
@dupondje but what do you do as part of the backup? do you call incremental backup?

**Comment by dupondje** [comment]
> @dupondje but what do you do as part of the backup? do you call incremental backup?

Yes using the incremental backup api.
But that doesn't matter, as we just call get a list of all the vm's first with their ovf before we loop over them and write the ovf.

It can even be simulated by just browsing to the api URL and then you have like 10-20 out of 400 that are missing the ovf configuration in the xml. And this is done while there are no other actions on the vm's.

**Comment by ahadas** [comment]
ok so that might explain it because the devices hash changes when we create a snapshot for incremental backup
would it be possible for you to query the OVF before initiating the backup?

**Comment by dupondje** [comment]
> ok so that might explain it because the devices hash changes when we create a snapshot for incremental backup would it be possible for you to query the OVF before initiating the backup?

Well that's what I mentioned above. Even if I just do the API call in a browser without any other actions/backups running, 10-20 out of 400 are missing the ovf data. So I think it has nothing to do with the backups itself.

**Comment by ahadas** [comment]
> > ok so that might explain it because the devices hash changes when we create a snapshot for incremental backup would it be possible for you to query the OVF before initiating the backup?
> 
> Well that's what I mentioned above. Even if I just do the API call in a browser without any other actions/backups running, 10-20 out of 400 are missing the ovf data. So I think it has nothing to do with the backups itself.

ack, so I'd start with that - because unmanaged devices shouldn't be removed unless there was some change to the VM. can you please file an issue with logs that demonstrate this?

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/GetVmOvfByVmIdQuery.java (hunk 1 above)
Comment (ahadas) [reply backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/GetVmOvfByVmIdQuery.java:44]: and also, we may get to this line with a vm that was queried with db-generation X from the database and right before getting here, the devices monitoring has already update the db-generation in the database to Y - so I think that this is not the way to go but even if someone is willing to block the monitoring thread like that, there need to be some changes here..

**Review by ahadas — CHANGES_REQUESTED** [review]

**Pull request closed without merging**